src/solvers/dgsem_p4est/dg_2d_manifold_in_3d_covariant.jl

```diff
@@ -0,0 +1,308 @@
+end
+
+# Evaluate the initial condition in spherical vector components, then transform to 
+# contravariant components for use as prognostic variables.
+function Trixi.compute_coefficients!(u, func, t, mesh::P4estMesh{2},
+                                     equations::AbstractCovariantEquations{2}, dg::DG,
+                                     cache)
+
+    # Store the auxiliary variables at the volume quadrature nodes in cache.elements
+    compute_auxiliary_variables!(cache.elements, mesh, equations, dg)
+
+    # Copy the appropriate interface values of the auxiliary variables into cache.interfaces
+    prolong2interfaces_auxiliary!(cache, mesh, dg)
+
+    for element in eachelement(dg, cache)
+        for j in eachnode(dg), i in eachnode(dg)
+            x_node = Trixi.get_node_coords(cache.elements.node_coordinates,
+                                           equations, dg, i, j, element)
+            aux_vars_node = get_node_aux_vars(cache.elements.auxiliary_variables,
+                                              equations, dg, i, j, element)
+            u_node = func(x_node, t, aux_vars_node, equations)
+            Trixi.set_node_vars!(u, u_node, equations, dg, i, j, element)
+        end
+    end
+end
+
+# Weak form kernel which uses contravariant flux components, passing the element container 
+# and node/element indices into the flux function to give the flux access to geometric 
+# quantities
+@inline function Trixi.weak_form_kernel!(du, u, element, mesh::P4estMesh{2},
+                                         nonconservative_terms::False,
+                                         equations::AbstractCovariantEquations{2},
+                                         dg::DGSEM, cache, alpha = true)
+    (; derivative_dhat) = dg.basis
+
+    for j in eachnode(dg), i in eachnode(dg)
+        u_node = Trixi.get_node_vars(u, equations, dg, i, j, element)
+
+        aux_vars_node = get_node_aux_vars(cache.elements.auxiliary_variables,
+                                          equations, dg, i, j, element)
+        contravariant_flux1 = flux(u_node, aux_vars_node, 1, equations)
+        contravariant_flux2 = flux(u_node, aux_vars_node, 2, equations)
+
+        for ii in eachnode(dg)
+            Trixi.multiply_add_to_node_vars!(du, alpha * derivative_dhat[ii, i],
+                                             contravariant_flux1, equations, dg, ii, j,
+                                             element)
+        end
+
+        for jj in eachnode(dg)
+            Trixi.multiply_add_to_node_vars!(du, alpha * derivative_dhat[jj, j],
+                                             contravariant_flux2, equations, dg, i, jj,
+                                             element)
+        end
+    end
+
+    return nothing
+end
+
+# Flux differencing kernel which uses contravariant flux components, passing the element
+# container and node/element indices into the two-point volume flux function to give the 
+# flux access to geometric quantities
```

Comment:
```suggestion
# Flux differencing kernel which uses contravariant flux components
```

Reply:
Done in https://github.com/trixi-framework/TrixiAtmo.jl/blob/73cd425993772a5a9144fb7b4326028a21ceb7e7/src/solvers/dgsem_p4est/dg_2d_manifold_in_3d_covariant.jl#L63-L69